stdlib/atomic.mli

```diff
@@ -51,3 +53,106 @@ val incr : int t -> unit
 
 (** [decr r] atomically decrements the value of [r] by [1]. *)
 val decr : int t -> unit
+
+(** {1:examples Examples}
+
+    {2 Basic Thread Coordination}
+
+    A basic use case is to have global counters that are updated in a
+    thread-safe way, for example to keep some sorts of metrics
+    over IOs performed by the program. Another basic use case is to coordinate
+    the termination of threads in a given program, for example when one thread
+    finds an answer, or when the program is shut down by the user.
+
+    Here, for example, we're going to try to find a number whose hash
+    satisfies a basic property. To do that, we'll run multiple threads which
+    will try random numbers until they find one that works.
+
+    Of course the output below is a sample run and will change every time
+    the program is run.
+
+    {[
+    (* use for termination *)
+    let stop_all_threads = Atomic.make false
+
+    (* total number of individual attempts to find a number *)
+    let num_attempts = Atomic.make 0
+
+    (* find a number that satisfies [p], by... trying random numbers
+       until one fits. *)
+    let find_number_where (p:int -> bool) =
+      let rand = Random.State.make_self_init() in
+      while not (Atomic.get stop_all_threads) do
+
+        let n = Random.State.full_int rand max_int in
+        ignore (Atomic.fetch_and_add num_attempts 1 : int);
+
+        if p (Hashtbl.hash n) then (
+          Printf.printf "found %d (hash=%d)\n%!" n (Hashtbl.hash n);
+          Atomic.set stop_all_threads true; (* signal all threads to stop *)
+        )
+      done;;
+
+
+    (* run multiple domains to search for a [n] where [hash n <= 100] *)
+    let () =
+      let criterion n = n <= 100 in
+      let threads =
+        Array.init 8
```

Comment:
Instead of `8` you could use `Domain.recommended_domain_count` here. (This helps drive down the point that domains are not lightweight threads you can spawn as many as you want.)